projects/plugins/jetpack/modules/carousel/jetpack-carousel.php

```diff
@@ -756,7 +756,7 @@ class_exists( 'Jetpack_AMP_Support' )
 				$html = str_replace( '<div ', '<div ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' ", $html );
 				$html = str_replace( '<ul class="wp-block-gallery', '<ul ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"wp-block-gallery", $html );
 				$html = str_replace( '<ul class="blocks-gallery-grid', '<ul ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"blocks-gallery-grid", $html );
-				$html = str_replace( '<figure class="wp-block-gallery blocks-gallery-grid', '<figure ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"wp-block-gallery  blocks-gallery-grid", $html );
+				$html = preg_replace( '/\<figure class="(wp-block-gallery.*has-nested-images.*)"/', '<figure ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"$1\"", $html );
```

Comment:
This has been working for me in testing, but I'm wondering if this regex is slightly too greedy? This is a bit of a fabricated example, but the regex group will match across multiple tags if the post content includes a figure element with `wp-block-gallery` but not `has-nested-images`, and any other element that has `has-nested-images` in it:

![image](https://user-images.githubusercontent.com/14988353/134124635-394d9f24-f238-4ed6-b221-94837ee20c92.png)

Are there any situations where `wp-block-gallery blocks-gallery-grid` isn't the first set of classes? If not, we might be able to swap it out for the `str_replace`, but if the order does change sometimes, then maybe something like the following might fix the above issue?

```
\<figure class="(wp-block-gallery[^"].*?has-nested-images.*?)"
```

The above adds a negated set after the gallery class to check that it doesn't include `"` before `has-nested-images` and makes the checks lazy, which seemed to make my test strings in regexr.com happy.

Another potential challenge is that the way the block gap support currently works in Gutenberg is that [it will inject](https://github.com/wordpress/gutenberg/blob/f0ee75922094f4d96cd72947970da4f956af61dd/lib/block-supports/spacing.php#L133-L133) a `style `attribute before the `class` attribute if one doesn't exist in the markup. This would mean that the string `<figure class="wp-block-gallery"` wouldn't match, which will be a problem longer-term for this code that's adding data to the container 🤔

I wonder if longer-term if it'd be worth refactoring the Jetpack code in [check_content_for_blocks](https://github.com/Automattic/jetpack/blob/e063c73904b8da170716b48d804e619fd4e02b9f/projects/plugins/jetpack/modules/carousel/jetpack-carousel.php#L214-L214) to use the blocks' render callbacks to update the markup instead of the entire post content? Since we've parked the gallery block gap support for the time being, I don't think this is a blocker for this PR, just something for us to keep in mind when we do get back to implementing it.